language/testing-infra/transactional-test-runner/src/framework.rs

```diff
@@ -0,0 +1,402 @@
+    }
+
+    pub fn add(&mut self, named_addr_opt: Option<String>, module: CompiledModule) {
+        let id = module.self_id();
+        if let Some(named_addr) = named_addr_opt {
+            self.named_addresses.insert(id.clone(), named_addr);
+        }
+
+        let processed = ProcessedModule {
+            module,
+            interface_file: None,
+        };
+        self.modules.insert(id, processed);
+    }
+}
+
+fn compile_source_unit(
+    pre_compiled_deps: Option<&FullyCompiledProgram>,
+    deps: &[String],
+    path: String,
+) -> Result<CompiledUnit> {
+    use move_lang::PASS_COMPILATION;
+    let (mut files, comments_and_compiler_res) = move_lang::Compiler::new(&[path], deps)
+        .set_pre_compiled_lib_opt(pre_compiled_deps)
+        .run::<PASS_COMPILATION>()?;
+    let units_or_diags = comments_and_compiler_res
+        .map(|(_comments, move_compiler)| move_compiler.into_compiled_units());
+
+    match units_or_diags {
+        Err(diags) => {
+            if let Some(pcd) = pre_compiled_deps {
+                for (file_name, text) in &pcd.files {
+                    // TODO This is bad. Rethink this when errors are redone
+                    if !files.contains_key(file_name) {
+                        files.insert(*file_name, text.clone());
+                    }
+                }
+            }
+
+            let error_buffer = if read_bool_env_var(move_command_line_common::testing::PRETTY) {
+                move_lang::diagnostics::report_diagnostics_to_color_buffer(&files, diags)
+            } else {
+                move_lang::diagnostics::report_diagnostics_to_buffer(&files, diags)
+            };
+            Err(anyhow!(String::from_utf8(error_buffer).unwrap()))
+        }
+        Ok(mut units) => {
+            let len = units.len();
+            if len != 1 {
+                panic!("Invalid input. Expected 1 compiled unit but got {}", len)
+            }
+            Ok(units.pop().unwrap())
+        }
+    }
+}
+
+fn compile_ir_module<'a>(
+    deps: impl Iterator<Item = &'a CompiledModule>,
+    address: AccountAddress,
+    file_name: &str,
+) -> Result<CompiledModule> {
+    use compiler::Compiler as IRCompiler;
+    let code = std::fs::read_to_string(file_name).unwrap();
+    IRCompiler::new(address, deps.collect()).into_compiled_module(file_name, &code)
+}
+
+fn compile_ir_script<'a>(
+    deps: impl Iterator<Item = &'a CompiledModule>,
+    file_name: &str,
+) -> Result<CompiledScript> {
+    use compiler::Compiler as IRCompiler;
+    let code = std::fs::read_to_string(file_name).unwrap();
+    let (script, _) = IRCompiler::new(/* unused */ AccountAddress::ZERO, deps.collect())
+        .into_compiled_script_and_source_map(file_name.into(), &code)?;
+    Ok(script)
+}
+
+pub fn run_test_impl<'a, Command, Adaptor, Split>(
+    path: &Path,
+    adaptor: &mut Adaptor,
+    split_command: Split,
+) -> Result<(), Box<dyn std::error::Error>>
```

Comment:
Why not just use `anayhow::Result` here?

Reply:
Tried changing this, and remembered it is so datatest_stable::harness! was happy